lib/BasicResolver.php

```diff
@@ -151,16 +151,21 @@ public function resolve(string $name, int $typeRestriction = null): Promise
                         }
                     }
                 } catch (NoRecordException $e) {
-                    try {
-                        /** @var Record[] $cnameRecords */
-                        $cnameRecords = yield $this->query($name, Record::CNAME);
-                        $name = $cnameRecords[0]->getValue();
-                        continue;
-                    } catch (NoRecordException $e) {
-                        /** @var Record[] $dnameRecords */
-                        $dnameRecords = yield $this->query($name, Record::DNAME);
-                        $name = $dnameRecords[0]->getValue();
-                        continue;
+                    // Look for aliases only when we have at least one attempt left
+                    if ($redirects < 4) {
+                        try {
+                            /** @var Record[] $cnameRecords */
+                            $cnameRecords = yield $this->query($name, Record::CNAME);
+                            $name = $cnameRecords[0]->getValue();
+                            continue;
+                        } catch (NoRecordException $e) {
+                            /** @var Record[] $dnameRecords */
+                            $dnameRecords = yield $this->query($name, Record::DNAME);
+                            $name = $dnameRecords[0]->getValue();
+                            continue;
+                        }
+                    } else {
+                        throw $e;
```

Comment:
I think we should throw an exception here that says the `CNAME` chain is too long.

Reply:
@kelunik I think it has to be NoRecordsException, though the message could be changed, do you agree?

Reply:
You mean for BC? If we stop following `CNAME` records, because that's the resolvers policy, then there are records, so a `NoRecordException` doesn't seem appropriate.

Reply:
Yes, for BC

Reply:
@kelunik I introduced `RedirectException`. What do you think about it?

Reply:
I decided to throw a generic `DnsException` for now.

Reply:
But this is a special case, if other `DnsException` exceptions are temporary and indicate problems on the dns server, here the exception occurs on bad records, for example when there is a cyclic aliasing with `CNAME` record. I believe it should have its own class.

Reply:
The good thing about throwing a `DnsException` is that we can always sub-type it later if someone needs that differentiation. Do you need that currently?

Reply:
I do :)

Reply:
@kelunik could use please subtype it, when you have time?